lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -0,0 +1,631 @@
+  // Find the range of instructions that have to be analyzed in order to
+  // optimize the given os log call. The relevant instructions begin with the
+  // `allocStack` of the string interpolation type passed to the log call
+  // (namely, OSLogMessage) and end at the last instruction of the body of the
+  // called os log function (which will be inlined).
+  auto *stringInterpolAllocInst =
+      findOSLogMessageArgument(oslogCall)->getDefiningInstruction();
+  if (!stringInterpolAllocInst ||
+      !isa<AllocStackInst>(stringInterpolAllocInst)) {
+    diagnose(caller->getASTContext(), oslogCall->getLoc().getSourceLoc(),
+             diag::oslog_dynamic_message);
+    return;
+  }
+
+  auto firstI = findIteratorForInstruction(stringInterpolAllocInst);
+
+  // Load and link the called os log function before inlining. This is needed
+  // to link shared functions that are used in the callee body.
+  auto *callee = oslogCall->getReferencedFunction();
+  assert(callee);
+
+  if (callee->isExternalDeclaration()) {
+    callee->getModule().loadFunction(callee);
+    assert(!callee->isExternalDeclaration());
+    caller->getModule().linkFunction(callee,
+                                     SILModule::LinkingMode::LinkNormal);
+  }
+
+  // Inline the log call into the caller and find the last instruction of the
+  // log call.
+  auto lastBB = inlineFunction(oslogCall, caller, *this).second;
+  auto lastI = lastBB->end();
+
+  // Interpret the instructions from first to last and fold all compile-time
+  // constants found during interpretation.
+  constantFold(firstI, lastI);
+}
+
+void OSLogOptimization::constantFold(SILBasicBlock::iterator first,
+                                     SILBasicBlock::iterator last) {
+
+  SILInstruction *firstInst = &(*first);
+
+  // A mapping from SILValues to their constant values.
+  llvm::SmallMapVector<SingleValueInstruction *, SymbolicValue, 4> constantMap;
+
+  // An allocator to store symbolic values constructed during interpretation.
+  SymbolicValueBumpAllocator allocator;
+  ConstExprStepEvaluator constantEvaluator(allocator, firstInst->getFunction());
+
+  auto errorInfo =
+      collectConstants(first, last, constantMap, constantEvaluator);
+
+  // At this point, the `OSLogMessage` instance should be mapped to a symbolic
+  // value in the interpreter state. Furthermore, its format string and
+  // interger-valued fields (other than `OSLogArguments`) must be constants.
+  // If this is not the case, it means the formatting options or privacy
+  // qualifiers provided by the user were not inferred as compile-time
+  // constants. Detect and diagnose this scenario.
+  assert(isa<SingleValueInstruction>(firstInst));
+  bool errorDetected = detectAndDiagnoseErrors(
+      errorInfo, dyn_cast<SingleValueInstruction>(firstInst),
+      constantEvaluator);
+  if (errorDetected) {
+    return;
+  }
+  // The necessary string information must have been initialized here.
```

Comment:
Is it always guaranteed that during collectConstants we see a call to makeUTF8?

Reply:
Yeah, that's right. We have an empty string literal used in the initializer of the this struct and will be seen during traversal.